python/pyspark/sql/connect/session.py

```diff
@@ -489,13 +489,31 @@ def createDataFrame(
 
     createDataFrame.__doc__ = PySparkSession.createDataFrame.__doc__
 
-    def sql(self, sqlQuery: str, args: Optional[Union[Dict[str, Any], List]] = None) -> "DataFrame":
-        cmd = SQL(sqlQuery, args)
-        data, properties = self.client.execute_command(cmd.command(self._client))
-        if "sql_command_result" in properties:
-            return DataFrame.withPlan(CachedRelation(properties["sql_command_result"]), self)
-        else:
-            return DataFrame.withPlan(SQL(sqlQuery, args), self)
+    def sql(
+        self,
+        sqlQuery: str,
+        args: Optional[Union[Dict[str, Any], List]] = None,
+        **kwargs: Any,
+    ) -> "DataFrame":
+
+        if len(kwargs) > 0:
+            from pyspark.sql.connect.sql_formatter import SQLStringFormatter
+
+            formatter = SQLStringFormatter(self)
+            sqlQuery = formatter.format(sqlQuery, **kwargs)
+
+        try:
+            cmd = SQL(sqlQuery, args)
+            data, properties = self.client.execute_command(cmd.command(self._client))
+            if "sql_command_result" in properties:
+                return DataFrame.withPlan(CachedRelation(properties["sql_command_result"]), self)
+            else:
+                return DataFrame.withPlan(SQL(sqlQuery, args), self)
+        finally:
+            if len(kwargs) > 0:
```

Comment:
there is a bug in temp view, see https://issues.apache.org/jira/browse/SPARK-44406

it seems non-trivial, so I just create a ticket for tracking

Reply:
Can we call `cache` or `persist` on the DataFrame, and drop the table for now? If that doesn't work, I am fine as is too but we would have to at least clean the temp views ..

Reply:
do you mean this?

1, create tmp view;
2, cache the tmp view;
3, in `finally`: uncache and drop

Reply:
it seems `cache` doesn't help:

```
In [1]: df = spark.createDataFrame([(1, 4), (2, 4), (3, 6)], ["A", "B"])

In [2]: df.cache()
Out[2]: DataFrame[A: bigint, B: bigint]

In [3]: df.createOrReplaceTempView("t")

In [4]: df2 = spark.sql("select * from t")

In [5]: spark.catalog.dropTempView("t")
Out[5]: True

In [6]: df2.show()
23/07/13 13:03:56 ERROR SparkConnectService: Error during: execute. UserId: ruifeng.zheng. SessionId: 547b8abe-9802-4bc0-9b1c-241107b2c66c.
org.apache.spark.sql.AnalysisException: [TABLE_OR_VIEW_NOT_FOUND] The table or view `t` cannot be found. Verify the spelling and correctness of the schema and catalog.
If you did not qualify the name with a schema, verify the current_schema() output, or qualify the name with the correct schema and catalog.
To tolerate the error on drop use DROP VIEW IF EXISTS or DROP TABLE IF EXISTS.; line 1 pos 14;
'Project [*]
+- 'UnresolvedRelation [t], [], false
```

Reply:
>  we would have to at least clean the temp views

or we don't support the Dataframe parameter for now?

Reply:
Strictly logically speaking the existing behavior is weird because it caches the analysis result even though the view is gone. IMHO this behavior is almost undefined but I understand that it's existing. 

The current behavior almost feels like reference counting :( 

TBH I think it's ok with the difference in behavior. My 2c

Reply:
Even if I cache the result dataframe `df2`, trigger an action on it, it still not work.
I think the reason is that, each time the `df2` is used, it just send a `UnresolvedRelation [t]` to the server. After we drop the temp view, it can no longer be resolved.

This PR will create extra temp views, if `kwargs` contains DataFrames vars.
On way I can image is to cache the resolved plan in the server side, by using 
https://github.com/apache/spark/blob/57bbb4c6e7309e0beaeb3618978dcaf75ce9a7fe/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala#L131-L150